locator/network_topology_strategy.cc

```diff
@@ -268,7 +281,52 @@ void network_topology_strategy::validate_options(const gms::feature_service&) co
 
 std::optional<std::unordered_set<sstring>> network_topology_strategy::recognized_options(const topology& topology) const {
     // We only allow datacenter names as options
-    return topology.get_datacenters();
+    auto opts = topology.get_datacenters();
+    opts.merge(recognized_tablet_options());
+    return opts;
+}
+
+effective_replication_map_ptr network_topology_strategy::make_replication_map(table_id table, token_metadata_ptr tm) const {
+    if (!uses_tablets()) {
+        on_internal_error(rslogger, format("make_replication_map() called for table {} but replication strategy not configured to use tablets", table));
+    }
+    return do_make_replication_map(table, shared_from_this(), std::move(tm), _rep_factor);
+}
+
+future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(schema_ptr s, token_metadata_ptr tm) const {
+    tablet_map tablets(get_initial_tablets());
+
+    // FIXME: Don't use tokens to distribute nodes.
+    // The following reuses the existing token-based algorithm used by NetworkTopologyStrategy.
+    assert(!tm->sorted_tokens().empty());
+    auto token_range = tm->ring_range(dht::minimum_token());
+
+    for (tablet_id tb = 0; tb < tablets.tablet_count(); ++tb) {
+        natural_endpoints_tracker tracker(*tm, _dc_rep_factor);
+
+        while (true) {
+            co_await coroutine::maybe_yield();
+            inet_address ep = *tm->get_endpoint(*token_range.begin());
+            token_range.drop_front();
+            if (token_range.begin() == token_range.end()) {
+                token_range = tm->ring_range(dht::minimum_token());
+            }
+            if (tracker.add_endpoint_and_check_if_done(ep)) {
+                break;
```

Comment:
Won't this loop forever if there are not enough nodes in one of the DCs?

Reply:
No, because `rf` for a dc is reduced in the tracker if there is not enough nodes (`std::min(rf, node_count)`):

```
data_center_endpoints(size_t rf, size_t rack_count, size_t node_count, endpoint_set& endpoints, endpoint_dc_rack_set& racks)
            : _endpoints(endpoints)
            , _racks(racks)
            // If there aren't enough nodes in this DC to fill the RF, the number of nodes is the effective RF.
            , _rf_left(std::min(rf, node_count))
      
```